sdk/vitest.config.ts

```diff
@@ -0,0 +1,12 @@
+import { defineConfig } from 'vitest/config';
+
+export default defineConfig({
+	test: {
+		typecheck: {
+			only: true,
+			enabled: true,
+			include: ['tests/**/*.test.ts'],
```

Comment:
I think I'd prefer sticking to the default file extensions for typecheck tests (`test-d.ts` [^1]), so we don't need to configure and could place these test files next to the actual files, like we do with the "normal" tests.

This way, I think we could completely omit the config file and just add `--typecheck` to the script.

[^1]: https://vitest.dev/guide/testing-types.html

Reply:
Just adding `--typecheck` doesnt work even tho the docs say it should.

Reply:
Hmm, I guess it would with the `test-d.ts` file extension. I'm not sure whether we should mix normal tests with typecheck tests, because of the differences outlined in https://vitest.dev/guide/testing-types.html.

Reply:
> Hmm, I guess it would with the test-d.ts file extension.

that is what i have tested

> im not sure whether we should mix normal tests with typecheck tests

im not mixing them the SDK uses type tests exclusively

Reply:
Just trying to find a sensible typecheck setup here (could also become a thing in other packages in the future).
Whenever possible, I prefer to follow the official documentation and use the default values. If not possible, then it should preferably be explained with comments. Otherwise there is the risk that we will no longer know in retrospect why it was configured this way, which will make later adjustments/additions more difficult.

> > Hmm, I guess it would with the test-d.ts file extension.
> 
> that is what i have tested before moving to this structure

It does work for me:

<img width="779" alt="Screenshot 2024-03-13 at 15 10 16" src="https://github.com/directus/directus/assets/5363448/12ea6be4-1f8e-4ab0-b360-4506d697a5cf">

> im not mixing them the SDK uses type tests exclusively

I do think so (`expect` = "normal" test, `assertType` = typecheck test):
<img width="483" src="https://github.com/directus/directus/assets/5363448/d71a635a-31d1-46d7-99a9-8fa73419e61f">

Reply:
Ah cool, so yeah, here's what I would prefer
- remove legacy test ✅ 
- rename the typecheck test file to `test-d.ts` (as per docs/default) and place it next to the actual file as we're doing with the other tests - unless there's a reason for having it in a separate `tests` folder resp. we explicitly decide to do so for typecheck tests
- move the corresponding `schema.ts` file into the `__utils__` folder like we have it in the`app` and `app` packages
- remove the vitest config file
- change the script to `"test": "vitest --watch=false --typecheck"`
- revert the `tsconfig.json` file change

Reply:
So quick summary on this one:
- The tests here won't be bound to a single file they're testing and thus should not be mixed in with the sourcecode
- I prefer explicit configuration over trying to set these settings in a long cli arg chain
- And finally i do not see the need to move off the default `*.test.ts` for tests because they happen to do a type assetion

Reply:
> - The tests here won't be bound to a single file they're testing and thus should not be mixed in with the sourcecode

Ah okay, it seems to do so now
<img width="1266" src="https://github.com/directus/directus/assets/5363448/39d0607b-74ff-472c-b3d4-37a50581fdc3">
but if that will change with upcoming tests, then it sounds good to have them in a separate tests folders 👍 

> - I prefer explicit configuration over trying to set these settings in a long cli arg chain

👍 

> - And finally i do not see the need to move off the default *.test.ts for tests because they happen to do a type assetion

- It would be more explicit (allows you to recognize the test type by its name)
- It would align with the official docs
- It would not require extra config

Reply:
> Ah okay, it seems to do so now

This initial low hanging fruit test yeah as i've only committed the tests that caught this specific error but there will be more split up feature based tests in upcoming PRs based on this for example testing relation types across `fields`, `sort` `filters`, `output` etc.

Reply:
> there are differences between normal tests and typecheck tests, so it would be better to differentiate using the explicit extension. I don't see a reason for not following the standard intended for this.

Correct there are differences however as we're only using type tests for the SDK there is nothing to differentiate. The `test-d.ts` naming suggestion in this experimental feature is not explicitly needed to prevent conflicts between existing unit tests and new type test.
Thus i dont see a reason to not follow the commonly accepted `*.test.ts` pattern for unit tests in the SDK packages across editors and tools.
If more people insist, sure we'll rename the files it of course does not functionally matter.

Reply:
As you correctly point out it does not functionally matter, so works for me for now 😇👍